include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort_one_wg.h

```diff
@@ -106,7 +106,12 @@ struct __subgroup_radix_sort
         {
             const uint16_t __idx = __wi * __block_size + __i;
             if (__idx < __n)
-                new (&__values[__i]) _ValueT(::std::move(__src[__idx]));
+            {
+                if constexpr (std::is_trivially_move_constructible_v<_ValueT>)
+                    __values[__i] = std::move(__src[__idx]);
```

Comment:
This check conceptually says that if something is `trivially_move_constructible`, call move assignment, while there is no guarantee that the move assignment is available. This is like check `foo()` for availability and then call `bar()`, which is nonsense

Reply:
Fixed

Reply:
I believe the comment raised by Ruslan and myself is not about the type of check made.
The question is why should we call move assignment where move constructor is expected. The only thing that is guaranteed if the type is `trivially_move_constructible` is that we can do a bitwise move (e.g. `std::memmove`) instead of calling the move constructor. Move assignment is not guaranteed to be presented even if the type is trivial (not only trivially movable.
And it is also should be checked if using bitwise move instead of directly calling move constructor is beneficial.

Reply:
We already have this approach in our code: https://github.com/oneapi-src/oneDPL/blob/8b9a2dd9a347c0255e4759fa93f6738e5d103340/include/oneapi/dpl/pstl/algorithm_impl.h#L1498

Reply:
I understand, and at the same time it seems like we have issues caused by this approach:
https://godbolt.org/z/6Pad9M8bK.
I guess we need to discuss it within the team and either do it correctly in all places or remove this unclear optimizations.